### PR TITLE
AddNewFeatureImgclassification_visionmodel

### PR DESCRIPTION
**Pull request by KonakanchiSwathi**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/KonakanchiSwathi/mlflow/pull/10041?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10041/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10041
```

</p>
</details>

### Related Issues/PRs

 #xxx

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [x] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [x] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by KonakanchiSwathi** [comment]
> @KonakanchiSwathi Thank you for the contribution! Could you fix the following issue(s)?
> 
> #### ⚠ DCO check
> The DCO check failed. Please sign off your commit(s) by following the instructions [here](https://github.com/mlflow/mlflow/runs/17907540788). See https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#sign-your-work for more details.
Can you help on this issue

File: mlflow/transformers/__init__.py [hunk 1]
```diff
@@ -1309,7 +1309,7 @@ def _should_add_pyfunc_to_model(pipeline) -> bool:
         "DocumentQuestionAnsweringPipeline",
         "ImageToTextPipeline",
         "VisualQuestionAnsweringPipeline",
-        "ImageClassificationPipeline",
+        #"ImageClassificationPipeline",
```
Comment (BenWilson2) [inline mlflow/transformers/__init__.py:1312]: This can be deleted. No need to comment out.

Comment (KonakanchiSwathi) [reply mlflow/transformers/__init__.py:1312]: ok

File: mlflow/transformers/__init__.py [hunk 2]
```diff
@@ -1319,7 +1319,7 @@ def _should_add_pyfunc_to_model(pipeline) -> bool:
         "ZeroShotAudioClassificationPipeline",
     ]
 
-    impermissible_attrs = {"image_processor"}
+    impermissible_attrs = {""}
```
Comment (BenWilson2) [inline mlflow/transformers/__init__.py:1322]: 1322 to 1326 can be deleted as it's a no-op.

File: mlflow/transformers/__init__.py (hunk 1 above)
Comment (KonakanchiSwathi) [reply mlflow/transformers/__init__.py]: done

File: mlflow/transformers/__init__.py (hunk 2 above)
Comment (KonakanchiSwathi) [reply mlflow/transformers/__init__.py]: deleted

**Review by KonakanchiSwathi — COMMENTED** [review]
updated the changes and commented

File: mlflow/transformers/__init__.py (hunk 2 above)
Comment (KonakanchiSwathi) [reply mlflow/transformers/__init__.py]: made chages as per requested

File: mlflow/transformers/__init__.py [hunk 3]
```diff
@@ -2637,7 +2651,124 @@ def is_uri(s):
...
+                img=[] 
+                for i in range(len(data)):
+                    encoded_image = list(data[i].values())[0]
+                    if isinstance(encoded_image, str):
+                            self._validate_image_input_uri_or_file(encoded_image)
+     
+                    img.append(decode_image(encoded_image))        
+                    #return decode_image(encoded_image)
+                           
+                return img
```
Comment (BenWilson2) [inline mlflow/transformers/__init__.py:2737]: Can we use iterators here instead?

```suggestion
if isinstance(data, list) and all(isinstance(element, dict) for element in data):
    images = []
    for item in data:
        encoded_image = next(iter(item.values()))
        if isinstance(encoded_image, str):
            self._validate_image_input_uri_or_file(encoded_image)
        images.append(decode_image(encoded_image))
    return images
```

File: mlflow/transformers/__init__.py [hunk 4]
```diff
@@ -2637,7 +2651,124 @@ def is_uri(s):
...
+    @staticmethod
+    def _validate_image_input_uri_or_file(input_str):
+            """
+            Validation of blob references to image files, if a string is input to the ``predict``
+            method, perform validation of the string contents by checking for a valid uri or
+            filesystem reference instead of surfacing the cryptic stack trace that is otherwise raised
+            for an invalid uri input.
+            """
+
+            def is_uri(s):
```
Comment (BenWilson2) [inline mlflow/transformers/__init__.py:2751]: This is replicated from `_validate_str_input_uri_or_file`. Can that private static method be adapted to be abstract instead of copying the logic?

File: mlflow/transformers/__init__.py [hunk 5]
```diff
@@ -2637,7 +2651,124 @@ def is_uri(s):
...
+                    return all([result.scheme, result.netloc])
+                except ValueError:
+                    return False
+
+            valid_uri = os.path.isfile(input_str) or is_uri(input_str)
+
+            if not valid_uri:
+                raise MlflowException(
+                    "An invalid string input was provided. String inputs to "
+                    "audio files must be either a file location or a uri.",
```
Comment (BenWilson2) [inline mlflow/transformers/__init__.py:2763]: These are image files, right? When making the existing validation static method abstract, please adjust the Exception message to reflect the abstraction.

File: mlflow/transformers/__init__.py (hunk 4 above)
Comment (KonakanchiSwathi) [reply mlflow/transformers/__init__.py]: updated

File: mlflow/transformers/__init__.py (hunk 5 above)
Comment (KonakanchiSwathi) [reply mlflow/transformers/__init__.py]: used the _validate_str_input_uri_or_file and made it comment for both audio and image files

File: tests/test_Imageclassification.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+
```
Comment (BenWilson2) [inline tests/test_Imageclassification.py:1]: The tests should be in https://github.com/mlflow/mlflow/blob/master/tests/transformers/test_transformers_model_export.py 
And will need to verify a number of elements, as with other existing tests for pipeline types in transformers:

1. save_model and load_model functionality
2. log_model (with registration) and load_model
3. signature inference with validation of both input_example, signature validation with pyfunc serving and scoring
4. pyfunc input types (pandas DF formatting types (records, split, inputs)) 
5. input type encodings for multiple image types to ensure that the provided encoding and decoding functions correctly
6. assertions of inference structure conformance between native pipeline loading and pyfunc inference. 
7. assurance that uri-based inference works as intended
8. validation of defined exceptions (invalid uri throws, for instance)

File: tests/test_Imageclassification.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import mlflow
+import os
+from mlflow import pyfunc
+from datasets import load_dataset
+from transformers import ViTImageProcessor,ViTForImageClassification
+def vision_model_save():
+    Model_name="facebook/deit-base-patch16-224"
```
Comment (BenWilson2) [inline tests/test_Imageclassification.py:8]:
```suggestion
    model_name="facebook/deit-base-patch16-224"
```
local variables should be lowercase

File: tests/test_Imageclassification.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+       mlflow.transformers.save_model(
+                transformers_model=components,
+                path=path,
+        )
+    return  path 
+def load_mlmodel():
+    path=vision_model_save()
+    
+    model = mlflow.pyfunc.load_model(path)
+    url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
```
Comment (BenWilson2) [inline tests/test_Imageclassification.py:24]: When validating the uri functionality, can you change this to use the same methodology that we do for audio transcription validation? 
For those tests, we actually saved an open source (free use, courtesy of NASA and the library of congress) audio file and saved it within https://github.com/mlflow/mlflow/tree/master/tests/datasets . We should put a free use (please verify the license on the image used) image (or two, of different image formats (.jpg and .png, just to be safe) in this directory so that the uri validation is local to GitHub. That way, if there are issues in any source system that contains a uri referenced image that we're using, we're not beholden to them hosting the image (in which any changes that the host has would become a failure vector for CI).

Comment (KonakanchiSwathi) [reply tests/test_Imageclassification.py]: Added test cases

File: tests/test_Imageclassification.py (hunk 6 above)
Comment (KonakanchiSwathi) [reply tests/test_Imageclassification.py:1]: Added testcases

File: tests/test_Imageclassification.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+
+import mlflow
+import os
+from mlflow import pyfunc
+from datasets import load_dataset
+from transformers import ViTImageProcessor,ViTForImageClassification
+def vision_model_save():
```
Comment (rjaincc) [inline tests/test_Imageclassification.py:7]: Empty line before this

File: tests/test_Imageclassification.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    if not (os.path.exists(path)):
+       image_processor = ViTImageProcessor.from_pretrained(model_name)
+       model = ViTForImageClassification.from_pretrained(model_name)
+       components={"model": model,"image_processor":image_processor}
+       mlflow.transformers.save_model(
+                transformers_model=components,
+                path=path,
+        )
+    return  path 
+def load_mlmodel():
```
Comment (rjaincc) [inline tests/test_Imageclassification.py:20]: Empty line before this

File: tests/transformers/test_transformers_model_export.py [hunk 11]
```diff
@@ -3470,7 +3570,55 @@ def test_save_model_card_with_non_utf_characters(tmp_path, model_name):
     assert txt == card_data.text
     data = yaml.safe_load(tmp_path.joinpath(_CARD_DATA_FILE_NAME).read_text())
     assert data == card_data.data.to_dict()
+def test_vision_pipeline_pyfunc_predict_with_kwargs(small_vision_model):
```
Comment (rjaincc) [inline tests/transformers/test_transformers_model_export.py:3573]: Empty line before this

File: tests/transformers/test_transformers_model_export.py [hunk 12]
```diff
@@ -2068,6 +2127,47 @@ def test_qa_pipeline_pyfunc_predict(small_qa_pipeline):
...
+        model_uri = mlflow.get_artifact_uri(artifact_path)
+
+    
+    image_file_path = "https://raw.githubusercontent.com/mlflow/mlflow/master/tests/datasets/cat.png"  # Replace with the actual image file path
+
+    
+    image_data = image = Image.open(image_file_path)
+
+   
+    inference_payload = json.dumps({
```
Comment (rjaincc) [inline tests/transformers/test_transformers_model_export.py:2149]: one new line should be enough

File: mlflow/transformers/__init__.py [hunk 13]
```diff
@@ -1880,6 +1884,11 @@ def _predict(self, data):
             (transformers.AudioClassificationPipeline, transformers.TextClassificationPipeline),
         ):
             return pd.DataFrame(raw_output)
+        elif isinstance(
+            self.pipeline,
+            (transformers.ViTForImageClassification, transformers.ImageClassificationPipeline),
+        ):
+            return raw_output #pd.DataFrame(raw_output)
```
[the remark this answers is not in the document]
Comment (KonakanchiSwathi) [reply mlflow/transformers/__init__.py]: Done the change

File: tests/transformers/test_transformers_model_export.py (hunk 12 above)
Comment (KonakanchiSwathi) [reply tests/transformers/test_transformers_model_export.py:2145]: Done the change

File: tests/transformers/test_transformers_model_export.py (hunk 11 above)
Comment (KonakanchiSwathi) [reply tests/transformers/test_transformers_model_export.py:3569]: Done the change

File: tests/test_Imageclassification.py (hunk 9 above)
Comment (KonakanchiSwathi) [reply tests/test_Imageclassification.py:8]: done

File: tests/test_Imageclassification.py (hunk 10 above)
Comment (KonakanchiSwathi) [reply tests/test_Imageclassification.py:22]: done

**Comment by KonakanchiSwathi** [comment]
@BenWilson2  can you please review

File: mlflow/transformers/__init__.py [hunk 14]
```diff
@@ -2670,11 +2679,99 @@ def is_uri(s):
         if not valid_uri:
             raise MlflowException(
                 "An invalid string input was provided. String inputs to "
-                "audio files must be either a file location or a uri.",
+                "audio files/image files must be either a file location or a uri.",
```
Comment (BenWilson2) [inline mlflow/transformers/__init__.py:2682]:
```suggestion
                "audio or image pipelines must be either a file location or a uri.",
```

File: mlflow/transformers/__init__.py [hunk 15]
```diff
@@ -2652,7 +2661,7 @@ def decode_audio(encoded):
     @staticmethod
     def _validate_str_input_uri_or_file(input_str):
         """
-        Validation of blob references to audio files, if a string is input to the ``predict``
+        Validation of blob references to audio  files/image files, if a string is input to the ``predict``
```
Comment (BenWilson2) [inline mlflow/transformers/__init__.py:2664]:
```suggestion
        Validation of blob references for audio and image transformers pipelines; if a string is input to the ``predict``
```

File: tests/test_Imageclassification.py (hunk 6 above)
Comment (BenWilson2) [reply tests/test_Imageclassification.py:1]: Can this file please be removed? The fixture can be moved to the main transformers test file. We don't need this file here.

File: tests/transformers/test_transformers_model_export.py [hunk 16]
```diff
@@ -656,7 +680,27 @@ def test_qa_model_save_model_for_task_and_card_inference(small_seq2seq_pipeline,
     assert flavor_config["task"] == "text-classification"
     assert flavor_config["source_model_name"] == "lordtt13/emo-mobilebert"
 
+def test_vision_model_save_and_override_card(small_vision_model, model_path):
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:683]: We don't need this test. The card validation is pipeline implementation agnostic. Please delete this.

File: tests/transformers/test_transformers_model_export.py [hunk 17]
```diff
@@ -679,18 +723,6 @@ def test_qa_model_save_and_override_card(small_qa_pipeline, model_path):
     assert card_data["language"] == "en"
     assert card_data["license"] == "bsd"
     # Validate inferred model card text
-    with model_path.joinpath("model_card.md").open() as file:
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:682]: Can you please revert this deletion?

File: tests/transformers/test_transformers_model_export.py [hunk 18]
```diff
@@ -1340,6 +1372,33 @@ def test_qa_pipeline_pyfunc_load_and_infer(small_qa_pipeline, model_path, infere
     assert isinstance(pd_inference, list)
     assert all(isinstance(element, str) for element in inference)
 
+def test_Vision_pipeline_pyfunc_load_and_infer(small_vision_model, model_path, inference_payload):
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:1375]:
```suggestion
def test_vision_pipeline_pyfunc_load_and_infer(small_vision_model, model_path, inference_payload):
```

Comment (BenWilson2) [reply tests/transformers/test_transformers_model_export.py:1375]: What is inference_payload here? It hasn't been parametrized (the test function) and there is no fixture (nor should there be based on the other definitions of this variable within scope of other tests in this suite). 
Please review this and test these locally.

File: tests/transformers/test_transformers_model_export.py [hunk 19]
```diff
@@ -1340,6 +1372,33 @@ def test_qa_pipeline_pyfunc_load_and_infer(small_qa_pipeline, model_path, infere
...
+    signature = infer_signature(
+        inference_payload,
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:1377]: once an appropriate inference_payload is defined and validated, please make sure to test all of the different ways of passing images with the two different provided formats that have been added to this PR, as well as a URL pointing to the raw github content link to those files to validate the uri-capable inference signature for these pipeline types.

File: tests/transformers/test_transformers_model_export.py [hunk 20]
```diff
@@ -1340,6 +1372,33 @@ def test_qa_pipeline_pyfunc_load_and_infer(small_qa_pipeline, model_path, infere
...
+        mlflow.transformers.generate_signature_output(small_vision_model, inference_payload),
+    )
+
+    mlflow.transformers.save_model(
+        transformers_model=small_vision_model,
+        path=model_path,
+        signature=signature,
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:1384]: Can an input example be added to ensure that this functionality behaves correctly with these data types (byte encoded image data)

File: tests/transformers/test_transformers_model_export.py [hunk 21]
```diff
@@ -2068,6 +2127,43 @@ def test_qa_pipeline_pyfunc_predict(small_qa_pipeline):
...
+    # Log the image classification model
+    with mlflow.start_run():
+        mlflow.transformers.log_model(
+            transformers_model=small_vision_model,
+            artifact_path=artifact_path,
+        )
+        model_uri = mlflow.get_artifact_uri(artifact_path)
+
+    
+    image_file_path = "https://raw.githubusercontent.com/mlflow/mlflow/master/tests/datasets/cat.png"  # Replace with the actual image file path
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:2143]: this isn't yet present in the master branch, right? Perhaps splitting this into two separate PRs (one for uploading the required image files so that they are in master prior to this PR executing to ensure that CI will validate this functionality will be best.

File: tests/transformers/test_transformers_model_export.py [hunk 22]
```diff
@@ -2068,6 +2127,43 @@ def test_qa_pipeline_pyfunc_predict(small_qa_pipeline):
...
+    inference_payload = json.dumps({
+        "inputs": {
+            "image": image_data  # Replace with the input field name for your image classification model
+        }
+    })
+
+   
+    response = pyfunc_serve_and_score_model(
+        model_uri,
+        data=inference_payload,
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:2154]: As mentioned above, we need to validate all of the different structures for serving inference for these data types. uri, list of uri, byte encoded image, list of byte encoded images, dict containing images, etc. 
For all supported structures, we need to have coverage to ensure that any changes to byte processing within pyfunc will remain compatible with this added pipeline type.

Comment (BenWilson2) [reply tests/transformers/test_transformers_model_export.py:2154]: parametrized approaches will be best here; please don't create new tests for each structure.

File: tests/transformers/test_transformers_model_export.py [hunk 23]
```diff
@@ -2068,6 +2127,43 @@ def test_qa_pipeline_pyfunc_predict(small_qa_pipeline):
...
+        content_type=pyfunc_scoring_server.CONTENT_TYPE_JSON,
+        extra_args=["--env-manager", "local"],
+    )
+
+    
+    predictions = PredictionsResponse.from_json(response.content.decode("utf-8")).get_predictions()
+
+   
+    expected_labels = ["tabby", "tabby cat"]  # Replace with your expected class labels
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:2163]: we don't need to validate that the model labels correctly; merely that we get non-empty strings as returned that. match the index count of submitted images or uri paths.

File: tests/transformers/test_transformers_model_export.py [hunk 24]
```diff
@@ -2068,6 +2127,43 @@ def test_qa_pipeline_pyfunc_predict(small_qa_pipeline):
...
+
+    
+    assert predictions == expected_labels
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:2166]: This effectively validates that we're processing the right data since we're comparing the raw output of the pipeline + model and ensuring that pyfunc is not mangling the inferred data output.

File: tests/transformers/test_transformers_model_export.py [hunk 25]
```diff
@@ -3470,7 +3566,55 @@ def test_save_model_card_with_non_utf_characters(tmp_path, model_name):
     assert txt == card_data.text
     data = yaml.safe_load(tmp_path.joinpath(_CARD_DATA_FILE_NAME).read_text())
     assert data == card_data.data.to_dict()
+def test_vision_pipeline_pyfunc_predict_with_kwargs(small_vision_model):
+    artifact_path = "image_classification_model"
+
+   
+    image_file_paths = ["https://raw.githubusercontent.com/mlflow/mlflow/master/tests/datasets/cat_image.jpg", "https://raw.githubusercontent.com/mlflow/mlflow/master/tests/datasets/tiger_cat.jpg"]  # Replace with actual image file paths
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:3573]: same note above re: splitting the PR so that these files actually exist in the location targeted here.

File: tests/transformers/test_transformers_model_export.py [hunk 26]
```diff
@@ -3470,7 +3566,55 @@ def test_save_model_card_with_non_utf_characters(tmp_path, model_name):
...
+    labels = ["tabby", "tabby cat", "tiger cat","Egyptian cat"]  # Replace with corresponding class labels
+
+    
+    inference_payload = json.dumps({
+        "inputs": {
+            "images": image_file_paths,  # Replace with the input field name for image classification
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:3579]: please remove these comments.

File: tests/transformers/test_transformers_model_export.py [hunk 27]
```diff
@@ -3470,7 +3566,55 @@ def test_save_model_card_with_non_utf_characters(tmp_path, model_name):
...
+        },
+        "params": {
+            "top_k": 2,  # Specify your image classification parameters here
+            "confidence_threshold": 0.8,  # Example parameter; adjust as needed
+        },
+    })
+
+    
+    expected_predictions = ["tabby", "tabby cat", "tiger cat","Egyptian cat"]  # Replace with expected class labels
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:3588]: please remove these comments, plus similar comment here for validating the actual output labels; we don't need to do that.

File: tests/transformers/test_transformers_model_export.py [hunk 28]
```diff
@@ -3470,7 +3566,55 @@ def test_save_model_card_with_non_utf_characters(tmp_path, model_name):
...
+
+    # Log the image classification model
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:3590]: no need for these comments.

File: tests/transformers/test_transformers_model_export.py [hunk 29]
```diff
@@ -3864,6 +4008,9 @@ def test_basic_model_with_accelerate_homogeneous_mapping_works(tmp_path):
 
     text = "Apples are delicious"
 
+
+    assert loaded(text) == pipeline(text)
```
Comment (BenWilson2) [inline tests/transformers/test_transformers_model_export.py:4012]: Why is this here? Can you revert this unrelated change that seems to be a typo?

File: mlflow/transformers/__init__.py [hunk 30]
```diff
@@ -1414,11 +1409,17 @@ def _get_default_pipeline_signature(pipeline, example=None, model_config=None) -
                 transformers.FillMaskPipeline,
                 transformers.TextGenerationPipeline,
                 transformers.Text2TextGenerationPipeline,
+                transformers.ImageClassificationPipeline,
```
Comment (serena-ruan) [inline mlflow/transformers/__init__.py:1412]: Should we remove this as we catch ImageClassificationPipeline in below elif?

File: mlflow/transformers/__init__.py [hunk 31]
```diff
@@ -1880,6 +1884,11 @@ def _predict(self, data):
             (transformers.AudioClassificationPipeline, transformers.TextClassificationPipeline),
         ):
             return pd.DataFrame(raw_output)
+        elif isinstance(
+            self.pipeline,
+            (transformers.ViTForImageClassification, transformers.ImageClassificationPipeline),
```
Comment (serena-ruan) [inline mlflow/transformers/__init__.py:1889]: Why do we have ViTForImageClassification here?

File: mlflow/transformers/__init__.py [hunk 32]
```diff
@@ -2670,11 +2679,99 @@ def is_uri(s):
...
             raise MlflowException(
                 "An invalid string input was provided. String inputs to "
-                "audio files must be either a file location or a uri.",
+                "audio files/image files must be either a file location or a uri.",
                 error_code=BAD_REQUEST,
             )
+    def _convert_image_input(self, data):
+            """
+            Conversion utility for decoding the base64 encoded bytes data of a raw imagefile when
+            parsed through model serving, if applicable. Direct usage of the pyfunc implementation
```
Comment (serena-ruan) [inline mlflow/transformers/__init__.py:2688]: This function looks 99% the same as _convert_audio_input. Can we extract the common functions out and refactor to make the function simpler?

File: mlflow/transformers/__init__.py [hunk 33]
```diff
@@ -1414,11 +1409,17 @@ def _get_default_pipeline_signature(pipeline, example=None, model_config=None) -
...
             ),
         ):
             return ModelSignature(
                 inputs=Schema([ColSpec("string")]), outputs=Schema([ColSpec("string")])
             )
+        elif isinstance(pipeline, transformers.ImageClassificationPipeline):
+            return ModelSignature(
+                inputs=Schema([ColSpec("string")]),
+                outputs=Schema([ColSpec("string", name="label")]),
+            )
```
Comment (serena-ruan) [inline mlflow/transformers/__init__.py:1422]: Can we add tests to ensure this is true for all data input formats we accept?

File: mlflow/transformers/__init__.py [hunk 34]
```diff
@@ -2670,11 +2679,99 @@ def is_uri(s):
         if not valid_uri:
             raise MlflowException(
                 "An invalid string input was provided. String inputs to "
-                "audio files must be either a file location or a uri.",
+                "audio or image pipelines must be either a file location or a uri.",
```
Comment (MadhuM02) [inline mlflow/transformers/__init__.py:2682]: "audio or image pipelines" -> "audio or image files"

File: mlflow/transformers/__init__.py [hunk 35]
```diff
@@ -1414,11 +1409,17 @@ def _get_default_pipeline_signature(pipeline, example=None, model_config=None) -
...
         ):
             return ModelSignature(
                 inputs=Schema([ColSpec("string")]), outputs=Schema([ColSpec("string")])
             )
+        elif isinstance(pipeline, transformers.ImageClassificationPipeline):
+            return ModelSignature(
+                inputs=Schema([ColSpec("string")]),
+                outputs=Schema([ColSpec("string", name="label")]),
+            )        
         elif isinstance(pipeline, transformers.TextClassificationPipeline):
```
Comment (MadhuM02) [inline mlflow/transformers/__init__.py:1423]: you can add Image classification pipeline here in this if segment

Comment (MadhuM02) [reply mlflow/transformers/__init__.py:1423]: and remove in above two if segments

File: mlflow/transformers/__init__.py [hunk 36]
```diff
@@ -1880,6 +1884,11 @@ def _predict(self, data):
             (transformers.AudioClassificationPipeline, transformers.TextClassificationPipeline),
         ):
             return pd.DataFrame(raw_output)
+        elif isinstance(
+            self.pipeline,
+            (transformers.ViTForImageClassification, transformers.ImageClassificationPipeline),
+        ):
+            return raw_output
```
Comment (MadhuM02) [inline mlflow/transformers/__init__.py:1891]: can you investigate if we can club this with textclassification if segment

File: tests/test_Imageclassification.py (hunk 6 above)
Comment (KonakanchiSwathi) [reply tests/test_Imageclassification.py:1]: done

File: tests/transformers/test_transformers_model_export.py (hunk 25 above)
Comment (harupy) [inline tests/transformers/test_transformers_model_export.py:3573]: can we use local paths?

File: mlflow/transformers/__init__.py (hunk 30 above)
Comment (serena-ruan) [reply mlflow/transformers/__init__.py:1412]: Please don't resolve the comment when it's not resolved :)

File: mlflow/transformers/__init__.py [hunk 37]
```diff
@@ -2670,10 +2678,55 @@ def is_uri(s):
...
+        This utility will convert this JSON encoded, base64 encoded text back into bytes for
+        input into the imageclassificationPipeline for inference.
+        """
+        if isinstance(data, list) and all(isinstance(element, dict) for element in data):
+            images = []
+            for item in data:
+                encoded_image = next(iter(item.values()))
+                if isinstance(encoded_image, str):
+                    self._validate_str_input_uri_or_file(encoded_image)
+                images.append(decode_audio(encoded_image))  # noqa: F821
```
Comment (serena-ruan) [inline mlflow/transformers/__init__.py:2723]: This won't work, `decode_audio` function is defined inside `_convert_audio_input`, could you extract that function out and rename it?

File: tests/transformers/test_transformers_model_export.py [hunk 38]
```diff
@@ -1341,6 +1366,81 @@ def test_qa_pipeline_pyfunc_load_and_infer(small_qa_pipeline, model_path, infere
...
+def raw_image_file(imagename):
+    datasets_path = pathlib.Path(__file__).resolve().parent.parent.joinpath("datasets")
+
+    return datasets_path.joinpath(imagename).read_bytes()
+
+
+@pytest.mark.parametrize(
+    "inference_payload",
+    [
+        "https://raw.githubusercontent.com/mlflow/mlflow/master/tests/datasets/cat.png",
```
Comment (serena-ruan) [inline tests/transformers/test_transformers_model_export.py:1378]: Could we use a variable for this since it appears so many times?

File: tests/transformers/test_transformers_model_export.py [hunk 39]
```diff
@@ -1341,6 +1366,81 @@ def test_qa_pipeline_pyfunc_load_and_infer(small_qa_pipeline, model_path, infere
...
+        raw_image_file("cat_image.jpg"),
+        [raw_image_file("cat_image.jpg"), raw_image_file("tiger_cat.jpg")],
+        json.dumps(
+            {
+                "inputs": {
+                    "image": base64.b64encode(raw_image_file("cat_image.jpg")).decode("ascii"),
+                },
+            }
+        ),
+        base64.b64encode(raw_image_file("cat_image.jpg")).decode("ascii"),
```
Comment (serena-ruan) [inline tests/transformers/test_transformers_model_export.py:1408]: Could you fix the failing tests? This becomes too long, we should convert it inside the test

File: mlflow/transformers/__init__.py [hunk 40]
```diff
@@ -2564,7 +2571,7 @@ def _convert_cast_lists_from_np_back_to_list(data):
                     parsed_data.append(entry)
             return parsed_data
 
-    def _convert_audio_input(self, data):
+    def _convert_audio_input(self, data,type="Audio"):
```
Comment (rjaincc) [inline mlflow/transformers/__init__.py:2574]:
```suggestion
    def _convert_audio_input(self, data, type="Audio"):
```

File: mlflow/transformers/__init__.py [hunk 41]
```diff
@@ -2641,18 +2648,28 @@ def decode_audio(encoded):
         # to return the only input format that the audio transcription pipeline permits:
         # a bytes input of a single element.
         if isinstance(data, list) and all(isinstance(element, dict) for element in data):
-            encoded_audio = list(data[0].values())[0]
-            if isinstance(encoded_audio, str):
-                self._validate_str_input_uri_or_file(encoded_audio)
-            return decode_audio(encoded_audio)
+            # if type=="Audio":
```
Comment (rjaincc) [inline mlflow/transformers/__init__.py:2651]: Let's please clean up this section of code.

File: mlflow/types/schema.py [hunk 42]
```diff
@@ -52,6 +53,8 @@ def __new__(cls, value, numpy_type, spark_type, pandas_type=None, python_type=No
         dt.date,
     )
     """64b datetime data."""
+    pilimage = (9, "PILImage", "PILImage", object, Image.Image)
+    """pil image."""
```
Comment (serena-ruan) [inline mlflow/types/schema.py:57]: Please revert this change, we shouldn't expect image to be one of the DataTypes, also we should not infer it directly as a type. You should pass string representing the path to the image, or the bytes, instead of the image itself.

Comment (MadhuM02) [reply mlflow/types/schema.py:57]: @serena-ruan  transformers imagepipelines accept PIL Image also as input. Wondering if we should also support it in mlflow models??

Comment (serena-ruan) [reply mlflow/types/schema.py:57]: I don't think we should support it in signatures. For general input, I feel like most cases users just pass in the path to the image, and you can always use bytes. If there're asks for this image type, we can discuss later :)

**Comment by KonakanchiSwathi** [comment]
@BenWilson2 @serena-ruan Please review

**Comment by KonakanchiSwathi** [comment]
closing it
